src/partition/repart.c

```diff
@@ -2235,17 +2235,27 @@ static int context_load_partition_table(Context *context) {
                 r = fdisk_save_user_sector_size(c, /* phy= */ 0, arg_sector_size);
         else {
                 uint32_t ssz;
+                struct stat st;
 
                 r = context_open_and_lock_backing_fd(context->node, arg_dry_run ? LOCK_SH : LOCK_EX,
                                                      &context->backing_fd);
                 if (r < 0)
                         return r;
 
+                if (fstat(context->backing_fd, &st) < 0)
+                        return log_error_errno(r, "Failed to stat %s: %m", context->node);
+
                 /* Auto-detect sector size if not specified. */
                 r = probe_sector_size_prefer_ioctl(context->backing_fd, &ssz);
                 if (r < 0)
                         return log_error_errno(r, "Failed to probe sector size of '%s': %m", context->node);
 
+                /* If we found the sector size and we're operating on a block the size, save it in
+                 * arg_sector_size so we can use it later to check if we're using the actual block device
+                 * sector size or are just using the offset of the GPT partition table as the sector size. */
+                if (r > 0 && S_ISBLK(st.st_mode))
+                        arg_sector_size = ssz;
```

Comment:
this will overwrite the command line value (if any), shouldn't it check that it's non-zero before overwriting, to let an explicit choice win?

Reply:
The command line value is zero if we're in this block

Reply:
ah I missed that, as it was just outside of the snippet shown here, never mind